python/pyarrow/dataset.py

```diff
@@ -53,3 +57,248 @@
+
+    Examples
+    --------
+
+    Specify the Schema for paths like "/2009/June":
+
+    >>> partitioning(pa.schema([("year", pa.int16()), ("month", pa.string())]))
+
+    or let the types be inferred by only specifying the field names:
+
+    >>> partitioning(["year", "month"])
+
+    For paths like "/2009/June", the year will be inferred as int32 while month
+    will be inferred as string.
+
+    Create a Hive scheme for a path like "/year=2009/month=11":
+
+    >>> partitioning(
+    ...     pa.schema([("year", pa.int16()), ("month", pa.int8())]),
+    ...     flavor="hive")
+
+    A Hive scheme can also be discovered from the directory structure (and
+    types will be inferred):
+
+    >>> partitioning(flavor="hive")
+
+    """
+    if flavor is None:
+        # default flavor
+        if isinstance(field_names, pa.Schema):
+            return DirectoryPartitioning(field_names)
+        elif isinstance(field_names, list):
+            return DirectoryPartitioning.discover(field_names)
+        elif field_names is None:
+            raise ValueError(
+                "For the default directory flavor, need to specify "
+                "'field_names' as Schema or list of field names")
+        else:
+            raise ValueError(
+                "Expected Schema or list of field names, got {0}".format(
+                    type(field_names)))
+    elif flavor == 'hive':
+        if isinstance(field_names, pa.Schema):
+            return HivePartitioning(field_names)
+        elif field_names is None:
+            return HivePartitioning.discover()
+        else:
+            raise ValueError(
+                "Expected Schema or None for 'field_names', got {0}".format(
+                    type(field_names)))
+    else:
+        raise ValueError("Unsupported flavor")
+
+
+def _ensure_fs_and_paths(path_or_paths, filesystem=None):
+    # Validate and convert the path-likes and filesystem.
+    # Returns filesystem and list of string paths or FileSelector
+    from pyarrow.fs import FileSystem, FileType, FileSelector
+
+    if isinstance(path_or_paths, list):
+        paths_or_selector = [_stringify_path(path) for path in path_or_paths]
+        if filesystem is None:
+            # infer from first path
+            filesystem, _ = FileSystem.from_uri(paths_or_selector[0])
+    else:
+        path = _stringify_path(path_or_paths)
+        if filesystem is None:
+            filesystem, path = FileSystem.from_uri(path)
+
+        stats = filesystem.get_target_stats([path])[0]
+        if stats.type == FileType.Directory:
+            # for directory, pass a selector
+            paths_or_selector = FileSelector(path, recursive=True)
+        elif stats.type == FileType.File:
+            # for a single file path, pass it as a list
+            paths_or_selector = [path]
+        else:
+            raise FileNotFoundError(path)
+
+    return filesystem, paths_or_selector
+
+
+def source(path_or_paths, filesystem=None, partition_scheme=None,
```

Comment:
I have a naming problem here. I wanted to rename this keyword now to `partitioning`, but that's not possible because we already have a function `partitioning()` defined above. So we will need to come up with a different name for one of both.

Maybe `partitions=` as keyword?

Reply:
@jorisvandenbossche I resolved this problem by https://github.com/kszucs/arrow/pull/14/files#diff-85a2de212ba6f06e71fbc143b6e9a543R135

Reply:
> I resolved this problem by introdicing another function

Done